common4j/src/main/com/microsoft/identity/common/java/cache/MsalOAuth2TokenCache.java

```diff
@@ -1092,16 +1091,45 @@ public AccountRecord getAccountByLocalAccountId(@Nullable final String environme
                                                     @NonNull final String clientId,
                                                     @NonNull final String localAccountId) {
         final String methodName = ":getAccountByLocalAccountId";
-
-        final List<AccountRecord> accounts = getAccounts(environment, clientId);
-
         Logger.verbosePII(
                 TAG + methodName,
                 "LocalAccountId: [" + localAccountId + "]"
         );
 
-        for (final AccountRecord account : accounts) {
-            if (localAccountId.equals(account.getLocalAccountId())) {
+        List<AccountRecord> accountsFilteredByEnvAndLocalAccountId = new ArrayList<>();
+        List<AccountRecord> accountRecordList = mAccountCredentialCache.getAccounts();
```

Comment:
>List<AccountRecord> accountRecordList = mAccountCredentialCache.getAccounts(); [](http://example.com/codeflow?start=8&length=78)

A couple more points on optimization

1. If accountRecordList size is zero, return early. This can be added in getAccounts(environment, client) method as well.
2. We are anyway calling getCredentialsFilteredBy() after getting accounts, so we can call that first to get appCredentials and then iterate over accounts

for (AccountRecord accountRecord: accountRecordList) {
            if (accountRecord.getLocalAccountId().equals(localAccountId)
                    && accountRecord.getEnvironment().equals(environment) && accountHasCredential(account, appCredentials)) {
               return accountRecord;
            }
        }